File: shared-ts/src/mqttUtils.ts
export const PUBLIC_OUT = 'public/out';
export const PUBLIC_IN = 'public/in';

export const picoOut = (module: string): string => `pico/${module}/out`;
export const picoIn = (module: string): string => `pico/${module}/in`;


export const reverseTopic = (topic: string): string => {
  if (topic === PUBLIC_OUT) return PUBLIC_IN;
  if (topic === PUBLIC_IN) return PUBLIC_OUT;

  const [, module, direction] = topic.split('/');
  return direction === 'out' ? picoIn(module) : picoOut(module);
};

export const extractModuleName = (topic: string): string => {
  const [channel, module] = topic.split('/');

  if (channel === 'public') return '';
  return module;
};

File: shared/mqttUtils.py
PUBLIC_OUT = 'public/out'
PUBLIC_IN = 'public/in'


def picoOut(module: bytes) -> str:
    return f"pico/{module.decode('utf-8')}/out"


def picoIn(module: bytes) -> str:
    return f"pico/{module.decode('utf-8')}/in"


def reverseTopic(topic: str) -> str:
    if topic == PUBLIC_OUT:
        return PUBLIC_IN
    elif topic == PUBLIC_IN:
        return PUBLIC_OUT

    [_, module, direction] = topic.split('/')
    return picoIn(module) if direction == 'out' else picoOut(module)


def extractModuleName(topic: str) -> str:
    [channel, module] = topic.split('/')

    if channel == 'public':
        return ''
    else:
        return module
